design-proposals/live-migration-for-bridged-pod-network.md

```diff
@@ -0,0 +1,54 @@
+# Overview
+This feature enables support of live-migrate for VMs with pod network in bridge mode.
+
+## Motivation
+Masquerade and slirp are not so performant compared to bridge mode.
+We want to use the less latency modes with the opportunity to live-migrate the virtual machines.
```

Comment:
I would be interested to understand the motivation to have a disruptive migration.
If such a migration can cause up to a minute of disruption, aren't there other means to recover a VM on a different node? I was under the impression that live-migration with high disruption of connectivity can cause more harm then good (applications will still run while disruption occurs, vs applications with the whole VM freeze and get restored).

Reply:
> If such a migration can cause up to a minute of disruption, aren't there other means to recover a VM on a different node?

From my tests this is happening within a few seconds not minutes, so even tcp sessions do not have time to break.

Reply:
how do tcp connections survive an ip address change?

Reply:
Some CNIs support specifying IP-address per pod. So in this way IP-address is not changed, but still requires to renew DHCP lease for updating routes inside the VM

Reply:
Did you try to use masquerade with multiqueue?
That's the result of a comparison we did in the past (masquerade binding, 8vcpu 8multiqueue)-

```
Masquerade overhead is minimized with multiqueue.
Masquerade vs. bridge binding
TCP Stream: -37% (nomq) -13% (mq)
UDP Stream:  -2% (nomq) -6% (mq)
TCP Latency: -24% (nomq) -3% (mq)
UDP Latency: -6% (nomq) -2% (mq)
```

Reply:
They won't survive. It will not survive even a link flickering... those TCP connections will just reset.
But this is the case with masquerade as well, so this is not something new.
I was more worried about the duration of the downtime (e.g. in the case of attach/detach).

Reply:
Well I just performed a test, I used the flowing commands:

server vm:
```
mkdir cgi-bin
printf '%s\n' '#!/bin/sh' 'cat /dev/urandom' > cgi-bin/1.sh
chmod +x cgi-bin/1.sh
python -m http.server --cgi
```

client vm:
```
curl 10.10.10.1:8000/cgi-bin/1.sh > /dev/null
```

Now I can assure that in case if IP address is not changed the tcp-connections are surviving


Also I checked flood ping in both cases. When MAC-address is changes and when is not, result:

**When MAC-address is the same (link down and link up interface):**

<details>
<summary>
47 packets lost
</summary>
<pre>
[fedora@vm1 ~]$ sudo ping -f 10.10.10.2
PING 10.10.10.2 (10.10.10.2) 56(84) bytes of data.
...............................................
--- 10.10.10.2 ping statistics ---
72544 packets transmitted, 72497 received, 0.0647883% packet loss, time 14943ms
rtt min/avg/max/mdev = 0.069/0.117/8.138/0.160 ms, ipg/ewma 0.205/0.106 ms
</pre>
</details>

**When MAC-address is changed (reattach network card):**  

<details>
<summary>
54 packets lost
</summary>
<pre>
[fedora@vm1 ~]$ sudo ping -f 10.10.10.2
PING 10.10.10.2 (10.10.10.2) 56(84) bytes of data.
......................................................
--- 10.10.10.2 ping statistics ---
102443 packets transmitted, 102389 received, 0.0527122% packet loss, time 20785ms
rtt min/avg/max/mdev = 0.069/0.116/39.293/0.222 ms, pipe 4, ipg/ewma 0.202/0.355 ms
</pre>
</details>